UserCode/CodeGPU/plugins/HeterogeneousHGCalEERecHitProducer.cc

```diff
@@ -0,0 +1,155 @@
+  vdata_.noise_fC          = ps.getParameter<edm::ParameterSet>("HGCEE_noise_fC").getParameter<std::vector<double> >("values");
+  vdata_.rcorr             = ps.getParameter< std::vector<double> >("rcorr");
+  vdata_.weights           = ps.getParameter< std::vector<double> >("weights");
+  cdata_.s_hgcEE_fCPerMIP_ = vdata_.fCPerMIP.size();
+  cdata_.s_hgcEE_cce_      = vdata_.cce.size();
+  cdata_.s_hgcEE_noise_fC_ = vdata_.noise_fC.size();
+  cdata_.s_rcorr_            = vdata_.rcorr.size();
+  cdata_.s_weights_ = vdata_.weights.size();
+  cdata_.hgceeUncalib2GeV_ = 1e-6 / cdata_.hgcEE_keV2DIGI_;
+  vdata_.waferTypeL = {0, 1, 2};//ddd_->retWaferTypeL(); if depends on geometry the allocation is tricky!
+  cdata_.s_waferTypeL_ = vdata_.waferTypeL.size();
+
+  begin = std::chrono::steady_clock::now();
+
+  tools_.reset(new hgcal::RecHitTools());
+  stride_ = ( (nhitsmax_-1)/32 + 1 ) * 32; //align to warp boundary
+
+  allocate_memory_();
+
+  convert_constant_data_(h_kcdata_);
+
+  produces<HGCeeRecHitCollection>(collection_name_);
+}
+
+HeterogeneousHGCalEERecHitProducer::~HeterogeneousHGCalEERecHitProducer()
+{
+  delete kmdata_;
+  delete h_kcdata_;
+  delete d_kcdata_;
+  delete old_soa_;
+  delete d_oldhits_;
+  delete d_newhits_;
+  delete d_newhits_final_;
+  delete h_newhits_;
+
+  end = std::chrono::steady_clock::now();
+  std::cout << "Time difference (heterogeneous) = " << std::chrono::duration_cast<std::chrono::milliseconds>(end - begin).count() << " [ms]" << std::endl;
+}
+
+void HeterogeneousHGCalEERecHitProducer::acquire(edm::Event const& event, edm::EventSetup const& setup, edm::WaitingTaskWithArenaHolder w) {
+  const cms::cuda::ScopedContextAcquire ctx{event.streamID(), std::move(w), ctxState_};
+  set_geometry_(setup);
+  event.getByToken(token_, handle_ee_);
+  const auto &hits_ee = *handle_ee_;
+
+  unsigned int nhits = hits_ee.size();
+  convert_collection_data_to_soa_(hits_ee, old_soa_, nhits);
+
+  kmdata_ = new KernelModifiableData<HGCUncalibratedRecHitSoA, HGCRecHitSoA>(nhitsmax_, stride_, old_soa_, d_oldhits_, d_newhits_, d_newhits_final_, h_newhits_);
+  KernelManagerHGCalRecHit kernel_manager(kmdata_);
+  kernel_manager.run_kernels(h_kcdata_, d_kcdata_);
+  new_soa_ = kernel_manager.get_output();
+
+  rechits_ = std::make_unique<HGCRecHitCollection>();
+  convert_soa_data_to_collection_(*rechits_, new_soa_, nhits);
+}
+
+void HeterogeneousHGCalEERecHitProducer::produce(edm::Event& event, const edm::EventSetup& setup)
+{
+  cms::cuda::ScopedContextProduce ctx{ctxState_}; //only for GPU to GPU producers
+  event.put(std::move(rechits_), collection_name_);
+}
+
+void HeterogeneousHGCalEERecHitProducer::allocate_memory_()
+{
+  old_soa_ = new HGCUncalibratedRecHitSoA();
+  d_oldhits_ = new HGCUncalibratedRecHitSoA();
+  d_newhits_ = new HGCUncalibratedRecHitSoA();
+  d_newhits_final_ = new HGCRecHitSoA();
+  h_newhits_ = new HGCRecHitSoA();
```

Comment:
maybe `nullptr`?

Reply:
The memory allocation is needed. Bear with me: it would not be required if I would use those pointers to point to the memory chunks allocated later by ```cms::cuda::make_host_unique()``` and similar, but I am instead pointing the that memory indirectly, *i.e.*, using pointers inside the pointers you highlighted. As such, if I do not allocate memory for them as in the highlighted code snippet, they simply do not exist.